numba/tests/test_stencils.py

```diff
@@ -365,12 +407,1272 @@ def test_impl(n):
+
+    # not sure this makes sense, what size array should be returned when both args are labelled as
+    # operating under standard_indexing? It probably ought to raise
+    def test_basic51(self):
+        """2 args, standard_indexing, no relidx"""
+        def kernel(a, b):
+            return a[0, 1] + b[0, 2]
+        a = np.arange(12.).reshape(3, 4)
+        b = np.arange(12.).reshape(3, 4)
+        self.check(kernel, a, b, options={'standard_indexing': ['a', 'b']}, expected_exception=ValueError)
+
+    def test_basic52(self):
+        """3 args, standard_indexing on middle arg """
+        def kernel(a, b, c):
+            return a[0, 1] + b[0, 1] + c[1, 2]
+        a = np.arange(12.).reshape(3, 4)
+        b = np.arange(4.).reshape(2, 2)
+        c = np.arange(12.).reshape(3, 4)
+        self.check(kernel, a, b, c, options={'standard_indexing': 'b'})
+
+    # this should fail parfors with a more informative error message
+    # @stencil and @njit do not raise but they should
+    def test_basic53(self):
+        """2 args, standard_indexing on variable that does not exist"""
+        def kernel(a, b):
+            return a[0, 1] + b[0, 2]
+        a = np.arange(12.).reshape(3, 4)
+        b = np.arange(12.).reshape(3, 4)
+        ex = self.exception_dict(pyStencil=ValueError, stencil=Exception, parfor=KeyError, njit=Exception)
+        self.check(kernel, a, b, options={'standard_indexing': 'c'}, expected_exception=ex)
+
+    def test_basic54(self):
+        """2 args, standard_indexing, index from var"""
+        def kernel(a, b):
+            t = 2
+            return a[0, 1] + b[0, t]
+        a = np.arange(12.).reshape(3, 4)
+        b = np.arange(12.).reshape(3, 4)
+        self.check(kernel, a, b, options={'standard_indexing': 'b'})
+        
+    def test_basic55(self):
+        """2 args, standard_indexing, index from more complex var"""
+        def kernel(a, b):
+            s = 1
+            t = 2 - s
+            return a[0, 1] + b[0, t]
+        a = np.arange(12.).reshape(3, 4)
+        b = np.arange(12.).reshape(3, 4)
+        self.check(kernel, a, b, options={'standard_indexing': 'b'})
+        
+    def test_basic56(self):
+        """2 args, standard_indexing, added complexity """
+        def kernel(a, b):
+            s = 1
+            acc = 0 
+            for k in b[0, :]:
+                acc += k
+            t = 2 - s - 1
+            return a[0, 1] + b[0, t] + acc
+        a = np.arange(12.).reshape(3, 4)
+        b = np.arange(12.).reshape(3, 4)
+        self.check(kernel, a, b, options={'standard_indexing': 'b'})
+        
+    def test_basic57(self):
+        """2 args, standard_indexing, split index operation """
+        def kernel(a, b):
+            c = b[0]
+            return a[0, 1] + c[1]
+        a = np.arange(12.).reshape(3, 4)
+        b = np.arange(12.).reshape(3, 4)
+        self.check(kernel, a, b, options={'standard_indexing': 'b'})
+
+    # parfors fails
+    def test_basic58(self):
```

Comment:
parfors fails to work on this

Reply:
For this case, the code we get for the stencil kernel has an arrayexpr in it.  We do some manipulations and then run this through the compilation pipeline again.  This was crashing in type inference because it didn't know how to infer the type of arrayexpr.  I added something to typeof_expr to support arrayexpr and this test now passes.  I don't know if the code I added is totally correct or not (I doubt it is) so this should really be reviewed by an expert.